README.md
# Retz Digdag Plugin [![CircleCI](https://circleci.com/gh/retz/retz-digdag-plugin.svg?style=svg)](https://circleci.com/gh/retz/retz-digdag-plugin)

This plugin helps you to managing [Retz](https://github.com/retz/retz) ( a job queuing and execution service for [Apache Mesos](http://mesos.apache.org) ) jobs on
[Digdag](https://github.com/treasure-data/digdag) workflow.

```yaml
timezone: Asia/Tokyo

_export:
  plugin:
    repositories:
      - http://retz-maven.s3.amazonaws.com/releases
    dependencies:
      - io.github.retz:retz-digdag-plugin:0.2.0
  retz:
    appname: default
    cpu: 1
    mem: 512m

+simple_app:
  retz_run>: /opt/app/simple-app.sh

+heavy_batch:
  retz_run>: /opt/app/heavy-batch.sh ${session_date}
  cpu: 8
  mem: 16g
  priority: 1
  timeout: 30

+learn:
  retz_run>: /opt/app/learn.py
  appname: learn-ubuntu
  gpu: 2
  mem: 4g
  env: [FOO=123, BAR=456]
```

## Requirement
### Digdag
- [Digdag](https://github.com/treasure-data/digdag) Version: 0.9.13+

### Retz

This Plugin Version | [Retz Server](https://github.com/retz/retz/blob/master/doc/getting-started-server.rst) Version
--- | ---
0.0.x | 0.2.x (0.2.9+)
0.1.x | 0.3.x
0.2.x | 0.4.x

## Usage

### Digdag configuration property

To use the plugin, define the following Retz client settings to digdag config file (e.g. `digdag.properties`):

- `retz.server.uri`
  - Defines Retz server location to send all requests.
- `retz.authentication`
  - Set `true` to enable authentication.
- `retz.access.key`
  - Defines user identity to send requests to servers with.
- `retz.access.secret`
  - Defines access secret to identify and authenticate a user.

```properties
retz.server.uri = http://10.0.0.1:9090
retz.authentication = true
retz.access.key = deadbeef
retz.access.secret = cafebabe
```

Please refer to [Retz Client documents](https://github.com/retz/retz/blob/master/doc/api.rst#client-configuration-file) for details.

### Workflow settings

Include the following plugin settings in your digadg workflow file (`*.dig`):

```yaml
_export:
  plugin:
    repositories:
      - http://retz-maven.s3.amazonaws.com/releases
    dependencies:
      - io.github.retz:retz-digdag-plugin:0.2.0
```

## Provided operator plugins

### retz_run>: Submitting a Retz job

Submit a job by using Retz Client WebAPI. This operator schedules a job to Retz server and waits for it finish either successfully or not. Please refer to [Retz Client documents](https://github.com/retz/retz/blob/master/doc/api.rst#client-cli-and-api) for details.

```yaml
+run:
  retz_run>: echo "hello, Retz!"
  appname: your-test-app
```

### Options

- `retz_run>:` COMMAND [ARGS...]
    - **(required)** Remote command to run
- `appname`: STRING
    - **(required)** Application name you loaded by `retz-client load-app`
- `cpu`: NUMBER
    - Number of CPU cores assigned to the job
    - default: `1`
- `mem`: STRING
    - Number of size of RAM(MB) assigned to the job
    - default: `32MB`
    - This option can specify a size unit (e.g. `64m`, `128GB`)
        - MiB: `m`|`mb`, GiB:`g`|`gb`, TiB:`t`|`tb` (case insensitive)
- `disk`: STRING
    - Amount of temporary disk space in MB which the job is going to use
    - default: `32MB`
    - This option can specify a size unit (e.g. `64m`, `128GB`)
        - MiB: `m`|`mb`, GiB:`g`|`gb`, TiB:`t`|`tb` (case insensitive)
- `ports`: NUMBER
    - Number of ports (up to 1000) required to the job
    - default: `0`
- `gpu`: NUMBER
    - Number of GPU cards assigned to the job
    - default: `0`
- `priority`: NUMBER
    - Job priority. Priority handling depends on server planner setting
    - default: `0`.
- `name`: STRING
    - Human readable job name
    - default: `session_id`#`attempt_id`..`task_name`
- `tags`: [ARRAY OF NAMES]
    - Mark the job with tags
- `env`:  [ARRAY OF "KEY=VALUE"]
    - List of environment values. `$HOME` and `$MESOS_*` are overwritten by Mesos executor
- `timeout`: NUMBER
    - Timeout in minutes. After timeout, the client tries to kill the job
    - `-1` or `0` for no timeout
    - default: `1440` (24 hours)
- `verbose`: BOOLEAN
    - Set `true` to display detailed processing information
    - default: `false`
- `client_config`: STRING **(deprecated)**
    - Configuration file path for Retz Client.
    - This option is deperecated and will be removed in a future version.

### Output parameters

- `retz.last_job_id`
    - The job id this task executed.

### Examples

```yaml
_export:
  retz:
    verbose: true
    tags: ["${session_local_time}","${task_name}"]

+run_asakusa_m3bp:
  retz_run>: /opt/asakusa/yaess/bin/yaess-batch.sh m3bp.app -A date=${session_date}
  appname: m3bp-app
  cpu: 32
  mem: 8192
  priority: 2
  env: [
    "ASAKUSA_HOME=/opt/asakusa",
    "ASAKUSA_M3BP_OPTS=-Xmx${Math.round(mem * 0.3)}m",
  ]
  timeout: 120
```

## License
* [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)



RELEASE-NOTES.md
# Retz Digdag Plugin release notes

## 0.2.0 (2017-09-13)

* Bump the following version:
  * Digdag : 0.9.16
  * Retz Client : 0.4.0
    * Attention: This version requires Retz Server 0.4.0+. See: https://github.com/retz/retz/releases/tag/0.4.0

## 0.1.0 (2017-08-18)

* Bump the following version:
  * Digdag : 0.9.15
  * Retz Client : 0.3.1
  * Gradle : 4.1

## 0.0.4 (2017-08-18)

* Add duration of Retz task to state params ([0458000](https://github.com/retz/retz-digdag-plugin/commit/0458000e88a0ac6e2c63158a8c9ef5194b6c47de))

## 0.0.3 (2017-07-31)

* Make polling interval configurable ([ebcb47e](https://github.com/retz/retz-digdag-plugin/commit/ebcb47ea4e85e2f6350029c5791edf4bdbffc7c8) for detail)
* Enable to set Retz client settings in system config ([ad53aff](https://github.com/retz/retz-digdag-plugin/commit/ad53affcb761d5b7af2b1eebe50f1bbd0b11ce13) for detail)

## 0.0.2 (2017-07-27)

* Improve retz_run> operator for non-blocking (async polling).
* Enable to specify maven publish repository for testing.

## 0.0.1 (2017-07-19)

* The first release of Retz Digdag Plugin.

src/main/java/io/github/retz/digdag/plugin/RetzPlugin.java
package io.github.retz.digdag.plugin;

import io.digdag.client.config.Config;
import io.digdag.spi.*;

import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

public class RetzPlugin implements Plugin {
    @Override
    public <T> Class<? extends T> getServiceProvider(Class<T> type) {
        if (type == OperatorProvider.class) {
            return RetzOperatorProvider.class.asSubclass(type);
        } else {
            return null;
        }
    }

    public static class RetzOperatorProvider implements OperatorProvider {

        @Inject
        CommandExecutor exec;
        @Inject
        CommandLogger cLog;
        @Inject
        Config systemConfig;

        @Override
        public List<OperatorFactory> get() {
            return Arrays.asList(
                    new RetzRunOperatorFactory(exec, cLog, systemConfig)
            );
        }
    }
}

src/main/java/io/github/retz/digdag/plugin/RetzRunApiOperator.java
package io.github.retz.digdag.plugin;

import com.google.common.base.Optional;
import com.google.common.base.Throwables;
import io.digdag.client.config.Config;
import io.digdag.client.config.ConfigElement;
import io.digdag.client.config.ConfigException;
import io.digdag.spi.CommandLogger;
import io.digdag.spi.OperatorContext;
import io.digdag.spi.TaskExecutionException;
import io.digdag.spi.TaskResult;
import io.digdag.util.BaseOperator;
import io.digdag.util.Durations;
import io.github.retz.cli.ClientCLIConfig;
import io.github.retz.cli.TimestampHelper;
import io.github.retz.protocol.GetFileResponse;
import io.github.retz.protocol.GetJobResponse;
import io.github.retz.protocol.Response;
import io.github.retz.protocol.ScheduleResponse;
import io.github.retz.protocol.data.Job;
import io.github.retz.protocol.exception.JobNotFoundException;
import io.github.retz.web.Client;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.text.SimpleDateFormat;
import java.text.ParseException;
import java.time.Duration;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Properties;

import static java.nio.charset.StandardCharsets.UTF_8;

public class RetzRunApiOperator extends BaseOperator {

    private static final Logger LOGGER = LoggerFactory.getLogger(RetzRunApiOperator.class);

    private final CommandLogger clog;
    private final RetzOperatorConfig config;

    RetzRunApiOperator(OperatorContext context, RetzOperatorConfig config, CommandLogger clog) {
        super(context);
        this.config = config;
        this.clog = clog;
    }

    private static final String STATE_JOB_ID = "jobId";
    private static final String STATE_JOB_STATE = "jobState";
    private static final String STATE_POLL_ITERATION = "pollIteration";
    private static final String STATE_OFFSET = "offset";
    private static final String STATE_RESULT_CODE = "result";
    private static final String STATE_DURATION = "duration";
    private static final String STATE_REASON = "reason";

    private static final int MAX_FETCH_FILE_LENGTH = 65536;

    @Override
    public TaskResult runTask() {

        Config state = request.getLastStateParams().deepCopy();
        TaskResult taskResult;

        Optional<Integer> maybeEcode = state.getOptional(STATE_RESULT_CODE, Integer.class);
        if (!maybeEcode.isPresent()) {
            try (Client webClient = createClient()) {
                Optional<Integer> maybeJobId = state.getOptional(STATE_JOB_ID, Integer.class);
                Job job;
                if (!maybeJobId.isPresent()) {
                    job = processSchedule(webClient, state);
                } else {
                    job = getJob(maybeJobId.get(), webClient);
                }
                state.set(STATE_JOB_STATE, job.state().toString());
                LOGGER.debug("job: {}", job);

                TaskExecutionException nextPolling = processGetFile(job, webClient, state);
                LOGGER.debug("next polling: {}", state);

                throw nextPolling;
            }
        } else {
            taskResult = processFinish(maybeEcode.get(), state);
        }
        return taskResult;
    }

    private TaskResult processFinish(int result, Config state) {
        if (result != 0) {
            throw new TaskExecutionException(String.format(
                    "retz_run: Job(id=%s) failed. " +
                            "| retz-info: state=%s, reason=%s " +
                            "| log: `digdag log %s %s`",
                    state.get(STATE_JOB_ID, String.class),
                    state.get(STATE_JOB_STATE, String.class), state.get(STATE_REASON, String.class),
                    request.getAttemptId(), request.getTaskName()));
        }

        TaskResult taskResult = TaskResult.empty(request);
        taskResult.getStoreParams()
                .getNestedOrSetEmpty(RetzOperatorConfig.KEY_CONFIG_ROOT)
                .set("last_job_id", state.get(STATE_JOB_ID, String.class));

        return taskResult;
    }

    private Job processSchedule(Client webClient, Config state) {
        Job job = createJob();
        Response res;
        try {
            res = webClient.schedule(job);
        } catch (IOException ex) {
            throw new RuntimeException("Failed to schedule Retz job", ex);
        }
        if (!(res instanceof ScheduleResponse)) {
            throw new TaskExecutionException(String.format(
                    "Failed to schedule Retz job: %s",
                    res.status()));
        }

        Job scheduled = ((ScheduleResponse) res).job();
        LOGGER.info("Job(id={}) scheduled: {}", scheduled.id(), scheduled.state());

        initializeTaskState(scheduled, state);

        return scheduled;
    }

    private void initializeTaskState(Job job, Config state) {
        state.set(STATE_JOB_ID, job.id());
        state.set(STATE_POLL_ITERATION, 0);
        state.set(STATE_OFFSET, 0L);
    }


    private TaskExecutionException processGetFile(Job job, Client webClient, Config state) {

        switch(job.state()) {
            case QUEUED:
                checkTimeout(job, webClient);
                return nextPolling(state);
            case STARTING:
            case STARTED:
                checkTimeout(job, webClient);
                getWholeFileByState(job, webClient, "stdout", state);
                return nextPolling(state);
            case FINISHED:
            case KILLED:
                getWholeFileByState(job, webClient, "stdout", state);
                if (config.getVerbose()) {
                    LOGGER.info("Job(id={}) finished to get stdout, will get stderr", job.id());
                }
                getWholeFile(job, webClient, "stderr", 0);
                return finishJob(job, state);
            default:
                throw new IllegalStateException("unexpected status: " + job.state());
        }
    }

    private void checkTimeout(Job job, Client webClient) {
        int timeout = config.getTimeout();
        if (timeout > 0) {
            Date scheduled;
            try {
                scheduled = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX").parse(job.scheduled());
            } catch (ParseException ex) {
                throw Throwables.propagate(ex);
            }
            Calendar limit = Calendar.getInstance();
            limit.setTime(scheduled);
            limit.add(Calendar.MINUTE, timeout);

            if (limit.compareTo(Calendar.getInstance()) < 0) {
                try {
                    webClient.kill(job.id());
                    throw new TaskExecutionException(String.format(
                            "Job(id=%s) has been killed due to timeout after %d minute(s)",
                            job.id(), timeout));
                } catch (IOException ex) {
                    throw new RuntimeException(String.format(
                            "Job(id=%s) failed to kill timedout job", job.id()), ex);
                }
            }
        }
    }

    private TaskExecutionException nextPolling(Config state) {
        int iteration = state.get(STATE_POLL_ITERATION, Integer.class);
        int interval = exponentialBackoffInterval(iteration);
        state.set(STATE_POLL_ITERATION, ++iteration);

        return TaskExecutionException.ofNextPolling(interval, ConfigElement.copyOf(state));
    }

    private void getWholeFileByState(Job job, Client webClient, String filename, Config state) {
            long offset = state.get(STATE_OFFSET, Long.class);
            long bytesRead = getWholeFile(job, webClient, filename, offset);

            state.set(STATE_OFFSET, offset + bytesRead);
            if (bytesRead != 0) {
                state.set(STATE_POLL_ITERATION, 0);
            }
    }

    private long getWholeFile(Job job, Client webClient, String filename, long offset) {
        try {
            return readFileUntilEmpty(webClient, job.id(), filename, offset, new CommandLoggerBridge(clog, System.out));
        } catch (IOException ex) {
            throw new RuntimeException(String.format(
                    "Job(id=%s) failed with unexpected error", job.id()), ex);

        }
    }

    private static long readFileUntilEmpty(Client c, int id, String filename, long offset, OutputStream out) throws IOException {
        long current = offset;

        while (true) {
            Response res = c.getFile(id, filename, current, MAX_FETCH_FILE_LENGTH);
            if (res instanceof GetFileResponse) {
                GetFileResponse getFileResponse = (GetFileResponse) res;

                // Check data
                if (getFileResponse.file().isPresent()) {
                    if (getFileResponse.file().get().data().isEmpty()) {
                        // All contents fetched
                        return current - offset;

                    } else {
                        byte[] data = getFileResponse.file().get().data().getBytes(UTF_8);
                        LOGGER.debug("Fetched data length={}, current={}", data.length, current);
                        out.write(data);
                        current = current + data.length;
                    }
                } else {
                    //LOG.info("{}: ,{}", filename, current);
                    return current - offset;
                }
            } else {
                LOGGER.error(res.status());
                throw new IOException(res.status());
            }
        }
    }

    private TaskExecutionException finishJob(Job job, Config state) {
        String duration = getDuration(job.started(), job.finished());

        LOGGER.info("Job(id={}) finished in {}. status: {}",
                job.id(), duration, job.state());

        state.set(STATE_RESULT_CODE, job.result());
        state.set(STATE_DURATION, duration);
        state.remove(STATE_POLL_ITERATION);
        state.remove(STATE_OFFSET);

        if (job.result() != 0) {
            state.set(STATE_REASON, job.reason());
        }

        return TaskExecutionException.ofNextPolling(0, ConfigElement.copyOf(state));
    }

    private String getDuration(String started, String finished) {
        long duration;
        try {
            if (started == null || finished == null) {
                duration = -1L;
            } else {
                duration = TimestampHelper.diffMillisec(finished, started);
            }
        } catch(ParseException ex) {
            throw Throwables.propagate(ex);
        }

        if (duration < 0) {
            return "-";
        } else if (duration < 1000) {
            return String.format("%ss", duration / 1000.0);
        } else {
            return Durations.formatDuration(Duration.ofMillis(duration));
        }
    }


    private Client createClient() {
        Optional<String> maybeServerUri = config.getServerUri();
        if (maybeServerUri.isPresent()) {
            URI serverUri;
            try {
                serverUri = new URI(maybeServerUri.get());
            } catch (URISyntaxException ex) {
                throw new ConfigException(String.format(
                        "Invalid '%s': %s", RetzOperatorConfig.KEY_SYSCONF_SERVER_URI, maybeServerUri.get()), ex);
            }
            Properties p = new Properties();
            if (config.getAuthenticationEnabled().isPresent()) {
                p.setProperty(RetzOperatorConfig.KEY_SYSCONF_AUTH_ENABLED, config.getAuthenticationEnabled().get());
            }
            if (config.getAccessKey().isPresent()) {
                p.setProperty(RetzOperatorConfig.KEY_SYSCONF_ACCESS_KEY, config.getAccessKey().get());
            }
            if (config.getAccessSecret().isPresent()) {
                p.setProperty(RetzOperatorConfig.KEY_SYSCONF_ACCESS_SECRET, config.getAccessSecret().get());
            }
            RetzOperatorConfig.ClientConfig clientConfig = new RetzOperatorConfig.ClientConfig(p);

            return Client.newBuilder(serverUri)
                    .setAuthenticator(clientConfig.getAuthenticator())
                    .checkCert(!clientConfig.insecure())
                    .setVerboseLog(config.getVerbose())
                    .build();
        } else if (config.getClientConfig(workspace).isPresent()) {
            ClientCLIConfig fileConfig;
            try {
                fileConfig = new ClientCLIConfig(config.getClientConfig(workspace).get());
            } catch (IOException | URISyntaxException ex) {
                throw Throwables.propagate(ex);
            }

            return Client.newBuilder(fileConfig.getUri())
                    .setAuthenticator(fileConfig.getAuthenticator())
                    .checkCert(!fileConfig.insecure())
                    .setVerboseLog(config.getVerbose())
                    .build();
        } else {
            throw new ConfigException(String.format(
                    "fail to initialize Retz client: set '%s' to digdag configuration property",
                    RetzOperatorConfig.KEY_SYSCONF_SERVER_URI));
        }
    }

    private Job createJob() {
        String appName = config.getAppName();
        String remoteCmd = config.getRemoteCommand();
        Properties envProps = config.getEnvProps();

        int cpu = config.getCpu();
        int mem = config.getMemory();
        int disk = config.getDisk();
        int gpu = config.getGpu();
        int ports = config.getPorts();
        int priority = config.getPriority();
        String name = config.getJobName();
        List<String> tags = config.getTags();

        boolean verbose = config.getVerbose();

        Job job = new Job(appName, remoteCmd, envProps, cpu, mem, disk, gpu, ports);
        job.setPriority(priority);
        job.setName(name);
        job.addTags(tags);

        if (verbose) {
            LOGGER.info("Job created: {}", job);
        }

        return job;
    }

    private Job getJob(int id, Client webClient) {
        Response res;
        try {
            res = webClient.getJob(id);
        } catch (IOException ex) {
            throw Throwables.propagate(ex);
        }

        if (res instanceof GetJobResponse) {
            GetJobResponse getJobResponse = (GetJobResponse) res;
            if (getJobResponse.job().isPresent()) {
                return getJobResponse.job().get();
            } else {
                throw Throwables.propagate(new JobNotFoundException(id));
            }
        } else {
            throw new TaskExecutionException(String.format(
                    "Job(id=%s) getJob received invalid response: %s",
                    id, res.status()));
        }
    }

    private int exponentialBackoffInterval(int iteration) {
        return Math.min(Math.max(config.getMinPollInterval(), (int)Math.pow(2, iteration)), config.getMaxPollInterval());
    }

    static class CommandLoggerBridge extends OutputStream {

        private static final ThreadLocal<byte[]> CACHE = ThreadLocal.withInitial(() -> new byte[1]);

        private final CommandLogger commandLogger;
        private final OutputStream out;

        CommandLoggerBridge(CommandLogger commandLogger, OutputStream out) {
            this.commandLogger = commandLogger;
            this.out = out;
        }

        @Override
        public void write(int b) throws IOException {
            byte[] bytes = CACHE.get();
            bytes[0] = (byte) b;
            write(bytes, 0, 1);
        }

        @Override
        public void write(byte[] b, int off, int len) throws IOException {
            commandLogger.copy(new ByteArrayInputStream(b, off, len), out);
        }
    }
}

src/main/java/io/github/retz/digdag/plugin/RetzOperatorConfig.java
package io.github.retz.digdag.plugin;

import com.google.common.base.Optional;
import com.google.common.collect.ImmutableList;
import io.digdag.client.config.Config;
import io.digdag.client.config.ConfigException;
import io.digdag.spi.TaskRequest;
import io.digdag.util.Workspace;
import io.github.retz.cli.FileConfiguration;
import io.github.retz.cli.SubCommand;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.text.MessageFormat;
import java.util.List;
import java.util.Locale;
import java.util.Properties;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class RetzOperatorConfig {

    private static final Logger LOGGER = LoggerFactory.getLogger(RetzOperatorConfig.class);

    static final String KEY_CONFIG_ROOT = "retz";

    static final String KEY_SYSCONF_SERVER_URI = KEY_CONFIG_ROOT + ".server.uri";
    static final String KEY_SYSCONF_AUTH_ENABLED = KEY_CONFIG_ROOT + ".authentication";
    static final String KEY_SYSCONF_ACCESS_KEY = KEY_CONFIG_ROOT + ".access.key";
    static final String KEY_SYSCONF_ACCESS_SECRET = KEY_CONFIG_ROOT + ".access.secret";

    private static final String KEY_SYSCONF_MIN_POLL_INTERVAL = KEY_CONFIG_ROOT + ".min-poll-interval";
    private static final String KEY_SYSCONF_MAX_POLL_INTERVAL = KEY_CONFIG_ROOT + ".max-poll-interval";
    private static final int DEFAULT_MIN_POLL_INTERVAL = 1;
    private static final int DEFAULT_MAX_POLL_INTERVAL = 20;

    private static final String DEFAULT_CLIENT_CMD = "/opt/retz-client/bin/retz-client";
    private static final int DEFAULT_CPU = 1;
    private static final String DEFAULT_MEM = "32MB";
    private static final String DEFAULT_DISK = "32MB";
    private static final int DEFAULT_GPU = 0;
    private static final int DEFAULT_PORTS = 0;
    private static final int DEFAULT_PRIORITY = 0;
    private static final int DEFAULT_TIMEOUT = 24 * 60;

    private static final Pattern RETZ_SIZE_PATTERN = Pattern.compile("(\\d+)(\\D*)");

    private static final int RETZ_NAME_MAX = 32;

    private static final String NAME_ELLIPSIS = "..";

    private final TaskRequest taskRequest;
    private final Config retzConfig;
    private final Config systemConfig;

    RetzOperatorConfig(TaskRequest taskRequest, Config systemConfig) {
        this.taskRequest = taskRequest;
        this.retzConfig = taskRequest.getConfig().mergeDefault(
                taskRequest.getConfig().getNestedOrGetEmpty(KEY_CONFIG_ROOT));
        this.systemConfig = systemConfig;

        // TODO interim fix for error log: "Parameter 'xx' is not used at task ..."
        dumpConfig();
    }

    Config getRetzConfig() {
        return retzConfig;
    }

    String getClientMode() {
        return retzConfig.get("client_mode", String.class, "api");
    }

    Optional<String> getClientConfig(Workspace workspace) {
        Optional<String> config = retzConfig.getOptional("client_config", String.class);
        if (config.isPresent()) {
            LOGGER.warn("'client_config' is deprecated. use '{}' on digdag configuration property instead.",
                    KEY_SYSCONF_SERVER_URI);
            if (!new File(config.get()).isAbsolute()) {
                config = Optional.of(workspace.getPath(config.get()).toString());
            }
        }
        return config;
    }

    String getAppName() {
        if (!retzConfig.has("appname")) {
            throw new ConfigException("retz: 'appname' config is required");
        }
        return retzConfig.get("appname", String.class);
    }

    String getRemoteCommand() {
        return retzConfig.get("_command", String.class);
    }

    Properties getEnvProps() {
        return SubCommand.parseKeyValuePairs(
                retzConfig.getListOrEmpty("env", String.class));
    }

    int getCpu() {
        return retzConfig.get("cpu", Integer.class, DEFAULT_CPU);
    }

    int getMemory() {
        String value = retzConfig.get("mem", String.class, DEFAULT_MEM);
        return convertMB(value);
    }

    int getDisk() {
        String value = retzConfig.get("disk", String.class, DEFAULT_DISK);
        return convertMB(value);
    }

    int getGpu() {
        return retzConfig.get("gpu", Integer.class, DEFAULT_GPU);
    }

    int getPorts() {
        int ports = retzConfig.get("ports", Integer.class, DEFAULT_PORTS);
        if (ports < 0 || 1000 < ports) {
            throw new ConfigException(String.format(
                    "retz: --ports must be within 0 to 1000: %s",
                    ports));
        }
        return ports;
    }

    int getPriority() {
        return retzConfig.get("priority", Integer.class, DEFAULT_PRIORITY);
    }

    String getJobName() {
        return retzConfig.get("name", String.class, generateDefaultJobName(taskRequest));
    }

    List<String> getTags() {
        return retzConfig.getListOrEmpty("tags", String.class);
    }

    int getTimeout() {
        return retzConfig.get("timeout", Integer.class, DEFAULT_TIMEOUT);
    }

    boolean getVerbose() {
        return retzConfig.get("verbose", Boolean.class, false);
    }

    Optional<String> getServerUri() {
        return systemConfig.getOptional(KEY_SYSCONF_SERVER_URI, String.class);
    }

    Optional<String> getAuthenticationEnabled() {
        return systemConfig.getOptional(KEY_SYSCONF_AUTH_ENABLED, String.class);
    }

    Optional<String> getAccessKey() {
        return systemConfig.getOptional(KEY_SYSCONF_ACCESS_KEY, String.class);
    }

    Optional<String> getAccessSecret() {
        return systemConfig.getOptional(KEY_SYSCONF_ACCESS_SECRET, String.class);
    }

    int getMinPollInterval() {
        return systemConfig.get(KEY_SYSCONF_MIN_POLL_INTERVAL, Integer.class, DEFAULT_MIN_POLL_INTERVAL);
    }

    int getMaxPollInterval() {
        return systemConfig.get(KEY_SYSCONF_MAX_POLL_INTERVAL, Integer.class, DEFAULT_MAX_POLL_INTERVAL);
    }

    private boolean getStdErr() {
        return retzConfig.get("stderr", Boolean.class, true);
    }

    void addClientCommand(ImmutableList.Builder<String> command) {
        File clientPath = new File(retzConfig.get("client_cmd", String.class, DEFAULT_CLIENT_CMD));
        command.add(clientPath.getAbsolutePath());
    }

    void addClientConfig(ImmutableList.Builder<String> command, Workspace workspace) {
        if (getClientConfig(workspace).isPresent()) {
            command.add("-C").add(getClientConfig(workspace).get());
        }
    }

    void addVerbose(ImmutableList.Builder<String> command) {
        boolean verbose = getVerbose();
        if (verbose) {
            command.add("-v");
        }
    }

    void addRunSubCommand(ImmutableList.Builder<String> command) {
        command.add("run");
    }

    void addAppName(ImmutableList.Builder<String> command) {
        command.add("-A").add(getAppName());
    }

    void addJobName(ImmutableList.Builder<String> command) {
        command.add("-N").add(getJobName());
    }

    void addCpu(ImmutableList.Builder<String> command) {
        if (retzConfig.has("cpu")) {
            command.add("--cpu").add(String.valueOf(getCpu()));
        }
    }

    void addMemory(ImmutableList.Builder<String> command) {
        if (retzConfig.has("mem")) {
            command.add("--mem").add(String.valueOf(getMemory()));
        }
    }

    void addDisk(ImmutableList.Builder<String> command) {
        if (retzConfig.has("disk")) {
            command.add("--disk").add(String.valueOf(getDisk()));
        }
    }

    void addPorts(ImmutableList.Builder<String> command) {
        if (retzConfig.has("ports")) {
            command.add("--ports").add(String.valueOf(getPorts()));
        }
    }

    void addGpu(ImmutableList.Builder<String> command) {
        if (retzConfig.has("gpu")) {
            command.add("--gpu").add(String.valueOf(getGpu()));
        }
    }

    void addPriority(ImmutableList.Builder<String> command) {
        if (retzConfig.has("priority")) {
            command.add("--prio").add(String.valueOf(getPriority()));
        }
    }

    void addStdErr(ImmutableList.Builder<String> command) {
        boolean stderr = getStdErr();
        if (stderr) {
            command.add("--stderr");
        }
    }

    void addTimeout(ImmutableList.Builder<String> command) {
        if (retzConfig.has("timeout")) {
            command.add("--timeout").add(String.valueOf(getTimeout()));
        }
    }

    void addEnv(ImmutableList.Builder<String> command) {
        if (retzConfig.has("env")) {
            List<String> envList = retzConfig.getList("env", String.class);
            for (String env : envList) {
                command.add("-E").add(env);
            }
        }
    }

    void addTags(ImmutableList.Builder<String> command) {
        if (retzConfig.has("tags")) {
            command.add("--tags").add(String.join(",", getTags()));
        }

    }

    void addRemoteCommand(ImmutableList.Builder<String> command) {
        command.add("-c").add(getRemoteCommand());
        // command.add("-c").add("-");
        // UserSecretTemplate.of(getRemoteCommand()).format(context.getSecrets());
    }

    private String generateDefaultJobName(TaskRequest request) {
        /*
         * prefix of default job name:
         *   <attempt-id>
         *
         * default job name:
         *   (1) <prefix-head>..
         *   (2) <prefix>
         *   (3) <prefix>:..<task-name-tail>
         *   (4) <prefix>:<task-name>
         */

        String prefix = String.format("%d", request.getAttemptId());
        String taskName = request.getTaskName();

        // (1) <prefix-head>..
        if (prefix.length() > RETZ_NAME_MAX) {
            return prefix.substring(0, RETZ_NAME_MAX - NAME_ELLIPSIS.length()) + NAME_ELLIPSIS;
        }

        // (2) <prefix>
        //   we avoid non-ASCII characters for Retz database
        if (prefix.length() > RETZ_NAME_MAX - 2 - NAME_ELLIPSIS.length()
                || taskName.chars().anyMatch(it -> it > 0x7f)) {
            return prefix;
        }

        String candidate = String.format("%s:%s", prefix, taskName);

        // (3) <prefix>:..<task-name-tail>
        if (candidate.length() > RETZ_NAME_MAX) {
            String prefixPlus = String.format("%s:%s", prefix, NAME_ELLIPSIS);
            String taskNameTail = taskName.substring(
                    taskName.length() - (RETZ_NAME_MAX - prefixPlus.length()), taskName.length());
            return String.format("%s%s", prefixPlus, taskNameTail);
        }

        // (4) <prefix>:<task-name>
        return candidate;
    }

    private Integer convertMB(String value) {
        final Matcher matcher = RETZ_SIZE_PATTERN.matcher(value);
        if (!matcher.matches()) {
            throw new ConfigException(MessageFormat.format(
                    "retz: Invalid size: {0}",
                    value));
        }
        int count = Integer.parseInt(matcher.group(1));
        String unit = matcher.group(2);
        switch (unit.toLowerCase(Locale.ENGLISH)) {
            case "":
            case "m":
            case "mb":
                return count;
            case "g":
            case "gb":
                return count * 1024;
            case "t":
            case "tb":
                return count * 1024 * 1024;
            default:
                throw new ConfigException(MessageFormat.format(
                        "retz: Invalid size (unsupported size unit): {0}",
                        value));
        }
    }

    private void dumpConfig() {
        LOGGER.debug(
                "client_mode:{}, client_config:{}, appname:{}, env:{}, cpu:{}, mem:{}, disk:{}, gpu:{}, ports:{}, priority:{}, name:{}, tags:{}, timeout:{}, verbose:{}, stderr:{}",
                retzConfig.getOptional("client_mode", String.class),
                retzConfig.getOptional("client_config", String.class),
                retzConfig.getOptional("appname", String.class),
                retzConfig.getListOrEmpty("env", String.class),
                retzConfig.getOptional("cpu", Integer.class),
                retzConfig.getOptional("mem", String.class),
                retzConfig.getOptional("disk", String.class),
                retzConfig.getOptional("gpu", Integer.class),
                retzConfig.getOptional("ports", Integer.class),
                retzConfig.getOptional("priority", Integer.class),
                retzConfig.getOptional("name", String.class),
                retzConfig.getListOrEmpty("tags", String.class),
                retzConfig.getOptional("timeout", Integer.class),
                retzConfig.getOptional("verbose", Boolean.class),
                retzConfig.getOptional("stderr", Boolean.class)
        );
    }

    static class ClientConfig extends FileConfiguration {

        ClientConfig(Properties p) {
            super(p);
        }
    }

}
